File: packages/sdk/java/packages/admin/src/main/java/dev/edgebase/sdk/admin/VectorizeClient.java
// EdgeBase Java SDK — VectorizeClient for user-defined Vectorize indexes.
// Note: Vectorize is Edge-only. In local/Docker, the server returns stub responses.
package dev.edgebase.sdk.admin;

import dev.edgebase.sdk.core.*;

import java.util.*;

/**
 * Client for a user-defined Vectorize index.
 *
 * <pre>{@code
 * admin.vector("embeddings").upsert(List.of(Map.of("id", "doc-1", "values", List.of(0.1, 0.2))));
 * List<Map<String, Object>> results = admin.vector("embeddings").search(List.of(0.1, 0.2), 10, null);
 * }</pre>
 */
public class VectorizeClient {
    private final HttpClient httpClient;
    private final String index;

    public VectorizeClient(HttpClient httpClient, String index) {
        this.httpClient = httpClient;
        this.index = index;
    }

    /** Insert or update vectors. Returns mutation result with ok, count, mutationId. */
    @SuppressWarnings("unchecked")
    public Map<String, Object> upsert(List<Map<String, Object>> vectors) {
        Object res = httpClient.post("/vectorize/" + index,
                Map.of("action", "upsert", "vectors", vectors));
        return res instanceof Map ? (Map<String, Object>) res : Collections.emptyMap();
    }

    /** Insert vectors (errors on duplicate ID — server returns 409). */
    @SuppressWarnings("unchecked")
    public Map<String, Object> insert(List<Map<String, Object>> vectors) {
        Object res = httpClient.post("/vectorize/" + index,
                Map.of("action", "insert", "vectors", vectors));
        return res instanceof Map ? (Map<String, Object>) res : Collections.emptyMap();
    }

    /** Search for similar vectors. */
    @SuppressWarnings("unchecked")
    public List<Map<String, Object>> search(List<Double> vector, int topK, Map<String, Object> filter) {
        return search(vector, topK, filter, null, null, null);
    }

    /** Search with default topK=10. */
    public List<Map<String, Object>> search(List<Double> vector) {
        return search(vector, 10, null, null, null, null);
    }

    /** Search for similar vectors with full options. */
    @SuppressWarnings("unchecked")
    public List<Map<String, Object>> search(
            List<Double> vector, int topK, Map<String, Object> filter,
            String namespace, Boolean returnValues, String returnMetadata) {
        Map<String, Object> body = new HashMap<>();
        body.put("action", "search");
        body.put("vector", vector);
        body.put("topK", topK);
        if (filter != null) body.put("filter", filter);
        if (namespace != null) body.put("namespace", namespace);
        if (returnValues != null) body.put("returnValues", returnValues);
        if (returnMetadata != null) body.put("returnMetadata", returnMetadata);
        Object res = httpClient.post("/vectorize/" + index, body);
        if (res instanceof Map) {
            Object matches = ((Map<String, Object>) res).get("matches");
            return matches instanceof List ? (List<Map<String, Object>>) matches : Collections.emptyList();
        }
        return Collections.emptyList();
    }

    /** Search by an existing vector's ID (Vectorize v2 only). */
    @SuppressWarnings("unchecked")
    public List<Map<String, Object>> queryById(String vectorId, int topK, Map<String, Object> filter) {
        return queryById(vectorId, topK, filter, null, null, null);
    }

    /** Search by an existing vector's ID with full options. */
    @SuppressWarnings("unchecked")
    public List<Map<String, Object>> queryById(
            String vectorId, int topK, Map<String, Object> filter,
            String namespace, Boolean returnValues, String returnMetadata) {
        Map<String, Object> body = new HashMap<>();
        body.put("action", "queryById");
        body.put("vectorId", vectorId);
        body.put("topK", topK);
        if (filter != null) body.put("filter", filter);
        if (namespace != null) body.put("namespace", namespace);
        if (returnValues != null) body.put("returnValues", returnValues);
        if (returnMetadata != null) body.put("returnMetadata", returnMetadata);
        Object res = httpClient.post("/vectorize/" + index, body);
        if (res instanceof Map) {
            Object matches = ((Map<String, Object>) res).get("matches");
            return matches instanceof List ? (List<Map<String, Object>>) matches : Collections.emptyList();
        }
        return Collections.emptyList();
    }

    /** Retrieve vectors by their IDs. */
    @SuppressWarnings("unchecked")
    public List<Map<String, Object>> getByIds(List<String> ids) {
        Object res = httpClient.post("/vectorize/" + index,
                Map.of("action", "getByIds", "ids", ids));
        if (res instanceof Map) {
            Object vectors = ((Map<String, Object>) res).get("vectors");
            return vectors instanceof List ? (List<Map<String, Object>>) vectors : Collections.emptyList();
        }
        return Collections.emptyList();
    }

    /** Delete vectors by IDs. Returns mutation result with ok, count, mutationId. */
    @SuppressWarnings("unchecked")
    public Map<String, Object> delete(List<String> ids) {
        Object res = httpClient.post("/vectorize/" + index,
                Map.of("action", "delete", "ids", ids));
        return res instanceof Map ? (Map<String, Object>) res : Collections.emptyMap();
    }

    /** Get index info (vector count, dimensions, metric). */
    @SuppressWarnings("unchecked")
    public Map<String, Object> describe() {
        Object res = httpClient.post("/vectorize/" + index,
                Map.of("action", "describe"));
        return res instanceof Map ? (Map<String, Object>) res : Collections.emptyMap();
    }
}

File: packages/sdk/kotlin/admin/src/main/kotlin/dev/edgebase/sdk/admin/VectorizeClient.kt
// VectorizeClient — Vectorize index access for server-side use.
// Note: Vectorize is Edge-only. In local/Docker, the server returns stub responses.
package dev.edgebase.sdk.admin

import dev.edgebase.sdk.core.*

/**
 * Client for a user-defined Vectorize index.
 *
 * ```kotlin
 * admin.vector("embeddings").upsert(listOf(mapOf("id" to "doc-1", "values" to listOf(0.1, 0.2))))
 * val results = admin.vector("embeddings").search(listOf(0.1, 0.2), topK = 5)
 * ```
 */
class VectorizeClient internal constructor(
    private val client: HttpClient,
    private val index: String
) {
    /** Insert or update vectors. Returns mutation result with ok, count, mutationId. */
    @Suppress("UNCHECKED_CAST")
    suspend fun upsert(vectors: List<Map<String, Any?>>): Map<String, Any?> {
        val res = client.post("/vectorize/$index", mapOf("action" to "upsert", "vectors" to vectors))
        return res as? Map<String, Any?> ?: emptyMap()
    }

    /** Insert vectors (errors on duplicate ID — server returns 409). */
    @Suppress("UNCHECKED_CAST")
    suspend fun insert(vectors: List<Map<String, Any?>>): Map<String, Any?> {
        val res = client.post("/vectorize/$index", mapOf("action" to "insert", "vectors" to vectors))
        return res as? Map<String, Any?> ?: emptyMap()
    }

    /** Search for similar vectors. */
    @Suppress("UNCHECKED_CAST")
    suspend fun search(
        vector: List<Double>,
        topK: Int = 10,
        filter: Map<String, Any?>? = null,
        namespace: String? = null,
        returnValues: Boolean? = null,
        returnMetadata: String? = null
    ): List<Map<String, Any?>> {
        val body = mutableMapOf<String, Any?>(
            "action" to "search",
            "vector" to vector,
            "topK" to topK
        )
        if (filter != null) body["filter"] = filter
        if (namespace != null) body["namespace"] = namespace
        if (returnValues != null) body["returnValues"] = returnValues
        if (returnMetadata != null) body["returnMetadata"] = returnMetadata
        val res = client.post("/vectorize/$index", body)
        val map = res as? Map<String, Any?> ?: return emptyList()
        return map["matches"] as? List<Map<String, Any?>> ?: emptyList()
    }

    /** Search by an existing vector's ID (Vectorize v2 only). */
    @Suppress("UNCHECKED_CAST")
    suspend fun queryById(
        vectorId: String,
        topK: Int = 10,
        filter: Map<String, Any?>? = null,
        namespace: String? = null,
        returnValues: Boolean? = null,
        returnMetadata: String? = null
    ): List<Map<String, Any?>> {
        val body = mutableMapOf<String, Any?>(
            "action" to "queryById",
            "vectorId" to vectorId,
            "topK" to topK
        )
        if (filter != null) body["filter"] = filter
        if (namespace != null) body["namespace"] = namespace
        if (returnValues != null) body["returnValues"] = returnValues
        if (returnMetadata != null) body["returnMetadata"] = returnMetadata
        val res = client.post("/vectorize/$index", body)
        val map = res as? Map<String, Any?> ?: return emptyList()
        return map["matches"] as? List<Map<String, Any?>> ?: emptyList()
    }

    /** Retrieve vectors by their IDs. */
    @Suppress("UNCHECKED_CAST")
    suspend fun getByIds(ids: List<String>): List<Map<String, Any?>> {
        val res = client.post("/vectorize/$index", mapOf("action" to "getByIds", "ids" to ids))
        val map = res as? Map<String, Any?> ?: return emptyList()
        return map["vectors"] as? List<Map<String, Any?>> ?: emptyList()
    }

    /** Delete vectors by IDs. Returns mutation result with ok, count, mutationId. */
    @Suppress("UNCHECKED_CAST")
    suspend fun delete(ids: List<String>): Map<String, Any?> {
        val res = client.post("/vectorize/$index", mapOf("action" to "delete", "ids" to ids))
        return res as? Map<String, Any?> ?: emptyMap()
    }

    /** Get index info (vector count, dimensions, metric). */
    @Suppress("UNCHECKED_CAST")
    suspend fun describe(): Map<String, Any?> {
        val res = client.post("/vectorize/$index", mapOf("action" to "describe"))
        return res as? Map<String, Any?> ?: emptyMap()
    }
}
